File: android/src/main/java/com/mobilecoin/mobilecoin_flutter/FfiMistysignAttestedSession.java
// Copyright (c) 2021-2026 MobileCoin. All rights reserved.

package com.mobilecoin.mobilecoin_flutter;

import androidx.annotation.Keep;
import androidx.annotation.NonNull;

import com.mobilecoin.lib.TrustedIdentities;
import com.mobilecoin.lib.exceptions.AttestationException;
import com.mobilecoin.lib.exceptions.InvalidUriException;

import java.util.List;
import java.util.Map;

import mistysign.MistysignAttestedSession;
import mistysign.MistysignAttestedSessionException;

/**
 * Bridges {@link MistysignAttestedSession} to the plugin's method channel,
 * mirroring <code>FfiMistysignAttestedSession</code> on iOS.
 * <p>
 * Trusted identities arrive as the channel maps built by the Dart
 * <code>MistysignMrEnclave</code> and <code>MistysignMrSigner</code> types,
 * with their advisories comma joined into a single string.
 */
@Keep
public final class FfiMistysignAttestedSession {

    private static final String MR_ENCLAVE_KEY = "mrEnclave";
    private static final String MR_SIGNER_KEY = "mrSigner";
    private static final String PRODUCT_ID_KEY = "productId";
    private static final String MINIMUM_SECURITY_VERSION_KEY = "minimumSecurityVersion";
    private static final String CONFIG_ADVISORIES_KEY = "configAdvisories";
    private static final String HARDENING_ADVISORIES_KEY = "hardeningAdvisories";

    private static final String ADVISORY_SEPARATOR = ",";

    private FfiMistysignAttestedSession() {
    }

    public static int create() throws MistysignAttestedSessionException {
        try {
            final MistysignAttestedSession session = MistysignAttestedSession.create();
            final int hashCode = session.hashCode();
            ObjectStorage.addObject(hashCode, session);
            return hashCode;
        } catch (InvalidUriException | AttestationException exception) {
            throw new MistysignAttestedSessionException(
                    MistysignAttestedSessionException.Code.ATTESTATION_FAILED,
                    "Unable to create the Mistysign session",
                    exception);
        }
    }

    @NonNull
    public static byte[] authBeginRequestData(final int objectId, @NonNull final String responderId)
            throws MistysignAttestedSessionException {
        return sessionFor(objectId).authBeginRequestData(responderId);
    }

    public static void authEnd(final int objectId,
                               @NonNull final byte[] authResponseData,
                               @NonNull final List<Map<String, Object>> mrEnclaves,
                               @NonNull final List<Map<String, Object>> mrSigners)
            throws MistysignAttestedSessionException {
        // Evidence is only accepted when it matches a supplied identity, so an
        // empty set can never attest. Rejecting it here keeps the failure
        // legible: TrustedIdentities does not report how many it holds, so the
        // session itself cannot tell an empty set from a populated one.
        if (mrEnclaves.isEmpty() && mrSigners.isEmpty()) {
            throw new MistysignAttestedSessionException(
                    MistysignAttestedSessionException.Code.NO_TRUSTED_IDENTITIES,
                    "Attestation named no trusted identities to verify against.");
        }

        sessionFor(objectId).authEnd(
                authResponseData,
                trustedIdentities(mrEnclaves, mrSigners));
    }

    @NonNull
    public static byte[] encrypt(final int objectId, @NonNull final byte[] plaintext)
            throws MistysignAttestedSessionException {
        return sessionFor(objectId).encrypt(plaintext);
    }

    @NonNull
    public static byte[] decrypt(final int objectId, @NonNull final byte[] messageData)
            throws MistysignAttestedSessionException {
        return sessionFor(objectId).decrypt(messageData);
    }

    public static void deattest(final int objectId) throws MistysignAttestedSessionException {
        sessionFor(objectId).deattest();
    }

    public static boolean isAttested(final int objectId)
            throws MistysignAttestedSessionException {
        return sessionFor(objectId).isAttested();
    }

    public static void destroy(final int objectId) {
        final Object session = ObjectStorage.objectForKey(objectId);
        if (session instanceof MistysignAttestedSession) {
            // Drop the attested state rather than waiting for finalization,
            // which is what holds the session's native handle.
            ((MistysignAttestedSession) session).deattest();
        }
        ObjectStorage.removeObject(objectId);
    }

    @NonNull
    private static MistysignAttestedSession sessionFor(final int objectId)
            throws MistysignAttestedSessionException {
        final Object session = ObjectStorage.objectForKey(objectId);
        if (!(session instanceof MistysignAttestedSession)) {
            throw new MistysignAttestedSessionException(
                    MistysignAttestedSessionException.Code.NOT_ATTESTED,
                    "No Mistysign session for id " + objectId
                            + ". It was never created, or has been destroyed.");
        }
        return (MistysignAttestedSession) session;
    }

    @NonNull
    private static TrustedIdentities trustedIdentities(
            @NonNull final List<Map<String, Object>> mrEnclaves,
            @NonNull final List<Map<String, Object>> mrSigners)
            throws MistysignAttestedSessionException {
        try {
            final TrustedIdentities identities = new TrustedIdentities();

            for (final Map<String, Object> entry : mrEnclaves) {
                identities.addMrEnclaveIdentity(
                        measurement(entry, MR_ENCLAVE_KEY),
                        advisories(entry, CONFIG_ADVISORIES_KEY),
                        advisories(entry, HARDENING_ADVISORIES_KEY));
            }

            for (final Map<String, Object> entry : mrSigners) {
                identities.addMrSignerIdentity(
                        measurement(entry, MR_SIGNER_KEY),
                        shortValue(entry, PRODUCT_ID_KEY),
                        shortValue(entry, MINIMUM_SECURITY_VERSION_KEY),
                        advisories(entry, CONFIG_ADVISORIES_KEY),
                        advisories(entry, HARDENING_ADVISORIES_KEY));
            }

            return identities;
        } catch (AttestationException exception) {
            throw new MistysignAttestedSessionException(
                    MistysignAttestedSessionException.Code.ATTESTATION_FAILED,
                    "Unable to build the trusted identities",
                    exception);
        }
    }

    /**
     * Decodes a hex measurement, rejecting anything malformed.
     * <p>
     * {@link com.mobilecoin.lib.util.Hex#toByteArray} maps non hex characters
     * to -1 rather than failing, which would turn a typo into a measurement
     * that simply never matches. Failing here says why.
     */
    @NonNull
    private static byte[] measurement(@NonNull final Map<String, Object> entry,
                                      @NonNull final String key)
            throws MistysignAttestedSessionException {
        final Object value = entry.get(key);
        if (!(value instanceof String)) {
            throw malformed("Trusted identity is missing '" + key + "'");
        }

        final String hex = (String) value;
        if (hex.isEmpty() || hex.length() % 2 != 0) {
            throw malformed("'" + key + "' is not a whole number of hex encoded bytes");
        }

        final byte[] measurement = new byte[hex.length() / 2];
        for (int i = 0; i < measurement.length; ++i) {
            final int high = Character.digit(hex.charAt(i * 2), 16);
            final int low = Character.digit(hex.charAt(i * 2 + 1), 16);
            if (high < 0 || low < 0) {
                throw malformed("'" + key + "' is not hex encoded");
            }
            measurement[i] = (byte) ((high << 4) + low);
        }
        return measurement;
    }

    private static short shortValue(@NonNull final Map<String, Object> entry,
                                    @NonNull final String key)
            throws MistysignAttestedSessionException {
        final Object value = entry.get(key);
        if (!(value instanceof Number)) {
            throw malformed("Trusted identity is missing '" + key + "'");
        }

        final int intValue = ((Number) value).intValue();
        if (intValue < 0 || intValue > 0xFFFF) {
            throw malformed("'" + key + "' does not fit in an unsigned 16 bit value: " + intValue);
        }
        return (short) intValue;
    }

    /**
     * Splits the comma joined advisories the Dart side sends.
     * <p>
     * An absent value means none, and so does an empty one: <code>"".split(",")</code>
     * would otherwise yield a single empty advisory. A present value has to be
     * the string the channel contract specifies, because reading no advisories
     * out of a malformed one would quietly narrow what the identity tolerates,
     * leaving an enclave that legitimately carries them unable to attest with
     * nothing to point at.
     */
    @NonNull
    private static String[] advisories(@NonNull final Map<String, Object> entry,
                                       @NonNull final String key)
            throws MistysignAttestedSessionException {
        final Object value = entry.get(key);
        if (value == null) {
            return new String[0];
        }
        if (!(value instanceof String)) {
            throw malformed("'" + key + "' is not a comma joined string: "
                    + value.getClass().getName());
        }

        final String joined = (String) value;
        return joined.isEmpty() ? new String[0] : joined.split(ADVISORY_SEPARATOR);
    }

    @NonNull
    private static MistysignAttestedSessionException malformed(@NonNull final String message) {
        return new MistysignAttestedSessionException(
                MistysignAttestedSessionException.Code.ATTESTATION_FAILED, message);
    }

}

File: ios/mobilecoin_flutter/Sources/mobilecoin_flutter/FfiMistysignAttestedSession.swift
// Copyright (c) 2021-2026 MobileCoin. All rights reserved.

import Flutter
import Foundation
import MobileCoin

// Identity parsing rejects a malformed entry rather than dropping it. Dropping
// one cannot weaken trust -- evidence is only accepted against the identities
// that survive -- but it makes a misconfiguration invisible, leaving a typo to
// present as an enclave that never matches. Matches the Android bridge.
private func malformedIdentity(_ reason: String) -> MistysignAttestedSessionError {
    // Reuses the SDK's error so the Dart-visible code comes from the one
    // mapping in `flutterCode` and cannot drift from it.
    .attestationFailed(reason)
}

/// Splits the comma joined advisories the Dart side sends. An absent or empty
/// value means none.
private func mistysignAdvisories(_ value: Any?) -> [String] {
    (value as? String)?.split(separator: ",").map(String.init) ?? []
}

private func mistysignMeasurement(
    _ entry: [String: Any],
    _ key: String
) throws -> Data {
    guard let hex = entry[key] as? String else {
        throw malformedIdentity("Trusted identity is missing '\(key)'")
    }
    // HexEncoding rejects an odd length and any non-hex character, so a typo
    // cannot become a measurement that silently never matches.
    guard let measurement = HexEncoding.data(fromHexEncodedString: hex) else {
        throw malformedIdentity("'\(key)' is not hex encoded: \(hex)")
    }
    return measurement
}

private func mistysignUInt16(
    _ entry: [String: Any],
    _ key: String
) throws -> UInt16 {
    guard let value = entry[key] as? Int else {
        throw malformedIdentity("Trusted identity is missing '\(key)'")
    }
    guard let narrowed = UInt16(exactly: value) else {
        throw malformedIdentity("'\(key)' does not fit in an unsigned 16 bit value: \(value)")
    }
    return narrowed
}

private func mistysignMrEnclaves(from entries: [[String: Any]]) throws -> [Attestation.MrEnclave] {
    try entries.map { entry in
        let mrEnclave = try mistysignMeasurement(entry, "mrEnclave")
        do {
            return try Attestation.MrEnclave.make(
                mrEnclave: mrEnclave,
                allowedConfigAdvisories: mistysignAdvisories(entry["configAdvisories"]),
                allowedHardeningAdvisories: mistysignAdvisories(entry["hardeningAdvisories"])
            ).get()
        } catch {
            throw malformedIdentity("Invalid mrEnclave identity: \(error)")
        }
    }
}

private func mistysignMrSigners(from entries: [[String: Any]]) throws -> [Attestation.MrSigner] {
    try entries.map { entry in
        let mrSigner = try mistysignMeasurement(entry, "mrSigner")
        let productId = try mistysignUInt16(entry, "productId")
        let minimumSecurityVersion = try mistysignUInt16(entry, "minimumSecurityVersion")
        do {
            return try Attestation.MrSigner.make(
                mrSigner: mrSigner,
                productId: productId,
                minimumSecurityVersion: minimumSecurityVersion,
                allowedConfigAdvisories: mistysignAdvisories(entry["configAdvisories"]),
                allowedHardeningAdvisories: mistysignAdvisories(entry["hardeningAdvisories"])
            ).get()
        } catch {
            throw malformedIdentity("Invalid mrSigner identity: \(error)")
        }
    }
}

private extension MistysignAttestedSessionError {
    // Distinct codes let the Dart side discriminate the failure kind.
    var flutterCode: String {
        switch self {
        case .notAttested:
            return "MISTYSIGN_NOT_ATTESTED"
        case .noTrustedIdentities:
            return "MISTYSIGN_NO_TRUSTED_IDENTITIES"
        case .attestationFailed:
            return "MISTYSIGN_ATTESTATION_FAILED"
        case .encryptionFailed:
            return "MISTYSIGN_ENCRYPTION_FAILED"
        case .decryptionFailed:
            return "MISTYSIGN_DECRYPTION_FAILED"
        }
    }
}

struct FfiMistysignAttestedSession {
    struct Create: Command {
        func execute(args: [String: Any], result: @escaping FlutterResult) throws {
            let session = MistysignAttestedSession()
            let hash = ObjectIdentifier(session).hashValue
            ObjectStorage.addObject(session, forKey: hash)
            result(hash)
        }
    }

    struct AuthBeginRequestData: Command {
        func execute(args: [String: Any], result: @escaping FlutterResult) throws {
            guard
                let id = args["id"] as? Int,
                let session = ObjectStorage.objectForKey(id) as? MistysignAttestedSession,
                let responderId = args["responderId"] as? String,
                // The SDK treats the handshake as infallible and traps the
                // process on an empty responder id, so it cannot be let past.
                !responderId.isEmpty
            else {
                throw PluginError.invalidArguments
            }

            result(session.authBeginRequestData(responderId: responderId))
        }
    }

    struct AuthEnd: Command {
        func execute(args: [String: Any], result: @escaping FlutterResult) throws {
            guard
                let id = args["id"] as? Int,
                let session = ObjectStorage.objectForKey(id) as? MistysignAttestedSession,
                let authResponseData = args["authResponseData"] as? FlutterStandardTypedData,
                let mrEnclaveEntries = args["mrEnclaves"] as? [[String: Any]],
                let mrSignerEntries = args["mrSigners"] as? [[String: Any]]
            else {
                throw PluginError.invalidArguments
            }

            do {
                let attestation = Attestation(
                    mrEnclaves: try mistysignMrEnclaves(from: mrEnclaveEntries),
                    mrSigners: try mistysignMrSigners(from: mrSignerEntries)
                )
                try session.authEnd(authResponseData: authResponseData.data, attestation: attestation)
                result(nil)
            } catch let error as MistysignAttestedSessionError {
                result(FlutterError(code: error.flutterCode, message: error.localizedDescription, details: nil))
            }
        }
    }

    struct Encrypt: Command {
        func execute(args: [String: Any], result: @escaping FlutterResult) throws {
            guard
                let id = args["id"] as? Int,
                let session = ObjectStorage.objectForKey(id) as? MistysignAttestedSession,
                let plaintext = args["plaintext"] as? FlutterStandardTypedData
            else {
                throw PluginError.invalidArguments
            }

            do {
                result(try session.encrypt(plaintext.data))
            } catch let error as MistysignAttestedSessionError {
                result(FlutterError(code: error.flutterCode, message: error.localizedDescription, details: nil))
            }
        }
    }

    struct Decrypt: Command {
        func execute(args: [String: Any], result: @escaping FlutterResult) throws {
            guard
                let id = args["id"] as? Int,
                let session = ObjectStorage.objectForKey(id) as? MistysignAttestedSession,
                let messageData = args["messageData"] as? FlutterStandardTypedData
            else {
                throw PluginError.invalidArguments
            }

            do {
                result(try session.decrypt(messageData.data))
            } catch let error as MistysignAttestedSessionError {
                result(FlutterError(code: error.flutterCode, message: error.localizedDescription, details: nil))
            }
        }
    }

    struct Deattest: Command {
        func execute(args: [String: Any], result: @escaping FlutterResult) throws {
            guard
                let id = args["id"] as? Int,
                let session = ObjectStorage.objectForKey(id) as? MistysignAttestedSession
            else {
                throw PluginError.invalidArguments
            }

            session.deattest()
            result(nil)
        }
    }

    struct IsAttested: Command {
        func execute(args: [String: Any], result: @escaping FlutterResult) throws {
            guard
                let id = args["id"] as? Int,
                let session = ObjectStorage.objectForKey(id) as? MistysignAttestedSession
            else {
                throw PluginError.invalidArguments
            }

            result(session.isAttested)
        }
    }

    struct Destroy: Command {
        func execute(args: [String: Any], result: @escaping FlutterResult) throws {
            guard let id = args["id"] as? Int else {
                throw PluginError.invalidArguments
            }

            try ObjectStorage.removeObject(forKey: id, ofType: MistysignAttestedSession.self)
            result(nil)
        }
    }
}
